resharper/src/resharper-unity/Psi/ShaderLab/Parsing/ShaderLab.psi

```diff
@@ -888,52 +1022,75 @@ scalarPropertyValue
   NUMERIC_LITERAL<SHADER_LAB_VALUE, Number>
 ;
 
-vectorPropertyValue
+colorPropertyValue
 :
-  LPAREN<LPAREN, LParen>
-  vectorPropertyValues
-  RPAREN<RPAREN, RParen>
-;
-
-// Private and error handling to allow resync on vectorPropertyValue::RPAREN
-private errorhandling vectorPropertyValues
-:
-  LIST numericValue<Number, Numbers> SEP COMMA<COMMA, Comma>
+  colorLiteral<SHADER_LAB_VALUE, Color>
 ;
 
 texturePropertyValue
 :
   STRING_LITERAL<SHADER_LAB_NAME, Name>
   (
-    LBRACE<LBRACE, LBrace>    
-    (
-      texturePropertyValueTextureDimension
-      | texturePropertyValueTexGen
-      | matrixCommand // Legacy
-      | LIGHTMAP_MODE_KEYWORD<SHADER_LAB_KEYWORD, LightmapModeKeyword>
-    )?
+    LBRACE<LBRACE, LBrace>
+    texEnvProperty<TEX_ENV_PROPERTY, TexEnvProperty>*
     errorTexturePropertyBlockValues
     RBRACE<RBRACE, RBrace>
   )?
 ;
 
-texturePropertyValueTextureDimension
+interface texEnvProperty
 :
-  TEXTURE_2D_KEYWORD
-  | TEXTURE_3D_KEYWORD
-  | CUBE_KEYWORD
+  textureDimensionTexEnvProperty
+  | texGenTexEnvProperty
+  | matrixTexEnvProperty
+  | lightmapTexEnvProperty
 ;
 
-texturePropertyValueTexGen // Legacy 5.0
+textureDimensionTexEnvProperty
 :
-  TEXGEN_KEYWORD
-  (
-    CUBE_REFLECT_KEYWORD
-    | CUBE_NORMAL_KEYWORD
-    | OBJECT_LINEAR_KEYWORD
-    | EYE_LINEAR_KEYWORD
-    | SPHERE_MAP_KEYWORD
-  )
+  textureDimensionKeyword
+;
+
+private textureDimensionKeyword
+:
+  TEXTURE_2D_KEYWORD<SHADER_LAB_KEYWORD, Keyword>
+  | TEXTURE_3D_KEYWORD<SHADER_LAB_KEYWORD, Keyword>
+  | CUBE_KEYWORD<SHADER_LAB_KEYWORD, Keyword>
+  | TEXTURE_2D_ARRAY_KEYWORD<SHADER_LAB_KEYWORD, Keyword>
+  | CUBE_ARRAY_KEYWORD<SHADER_LAB_KEYWORD, Keyword>
+  | RECT_KEYWORD<SHADER_LAB_KEYWORD, Keyword>
+  | ANY_KEYWORD<SHADER_LAB_KEYWORD, Keyword>
+;
+
+// TODO: Shader compiler error - obsolete
```

Comment:
Technically it should be a warning, not an error (as it is in Unity). Hopefully, the versions of Unity where they are still working are really old.

Reply:
I thought Unity did treat this as an error. But anyway, good point. It'll be worth investigating when these constructs became errors/warnings and only highlight them if we're targeting that version or above. We could perhaps make the severity of our "compiler" errors configurable, unlike say C#.